tempodb/encoding/v0/data_reader.go

```diff
@@ -38,16 +38,26 @@ func (r *dataReader) Read(ctx context.Context, records []common.Record, buffer [
 		length += record.Length
 	}
 
-	buffer = make([]byte, length)
+	// Reset/resize buffer
+	if cap(buffer) < int(length) {
+		buffer = make([]byte, length)
+	}
+	buffer = buffer[:length]
+
 	_, err := r.r.ReadAt(ctx, buffer, int64(start))
 	if err != nil {
 		return nil, nil, err
 	}
 
-	slicePages := make([][]byte, 0, len(records))
+	//Reset/resize buffer
+	if cap(pagesBuffer) < len(records) {
```

Comment:
does pagesBuffer have any performance impact? the "real" cost should be allocating the underlying bytes buffer.

pagesBuffer would just prevent us from reallocating a few slice structs?

Reply:
Right, the main savings are already present in `buffer`.  However, in benchmarks there is about a 3% performance increase in BackendSearchBlock.Search when reusing the pagesBuffer too. Thoughts?

Reply:
Surprised at the 3%. Under normal operation reusing the buffer should save MBs of allocation while the slices are only a few bytes. My thought was it would be negligible and not worth the code complexity, but if it's really makes iterating through a block 3% faster than i'm ok with it.

Reply:
To clarify:  reusing `buffer` is ~30% improvement.   Reusing `pagesBuffer` is an additional ~3%.